# UDACITY PROJECT 4 - WRANGLE AND ANALYZE DATA
## DOG TWITTER DATA ANALYSIS 
### *Jhonatan Nagasako*
#### *24-FEB-2021*

<hr size="5"/>

<a id='contents'></a>
# Table of Contents

<ul>
<li><a href="#intro">A. INTRODUCTION</a></li>
<li><a href="#scope">B. PROJECT MOTIVATION-SCOPE</a></li>
<li><a href="#gather">1. GATHERING DATA</a></li>
<li><a href="#assess">2. ASSESSING DATA</a></li>
<li><a href="#clean">3. CLEANING DATA</a></li>
<li><a href="#store">4. STORING AND ACTING ON WRANGLED DATA</a></li>
<li><a href="#report">5. REPORT-DISCUSSION-CONCLUSION</a></li>
<li><a href="#files">6. PROJECT FILES</a></li>
</ul>

<hr size="5"/>

<a id='intro'></a>
# A. INTRODUCTION

Real-world data rarely comes clean. Using Python and its libraries, data was gathered from a variety of sources and in a variety of formats, assess its quality and tidiness, then clean it. This is called data wrangling. Data wrangling efforts was documented in a Jupyter Notebook, which was then showcased  through analyses and visualizations using Python (and its libraries) and/or SQL.

The dataset that used for wrangling (and analyzing and visualizing) is the tweet archive of Twitter user [@dog_rates](https://twitter.com/dog_rates), also known as [WeRateDogs](https://en.wikipedia.org/wiki/WeRateDogs). WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because ["they're good dogs Brent."](https://knowyourmeme.com/memes/theyre-good-dogs-brent). WeRateDogs has over 4 million followers and has received international media coverage.

WeRateDogs [downloaded their Twitter archive](https://help.twitter.com/en/managing-your-account/how-to-download-your-twitter-archive) and sent it to Udacity via email exclusively for you to use in this project. This archive contains basic tweet data (tweet ID, timestamp, text, etc.) for all 5000+ of their tweets as they stood on August 1, 2017. More on this soon.


![dog and twitter](https://video.udacity-data.com/topher/2017/October/59dd378f_dog-rates-social/dog-rates-social.jpg)

*Image via [Boston Magazine](https://www.bostonmagazine.com/arts-entertainment/2017/04/18/dog-rates-mit/)*

<a href="#contents">[Table of Contents]</a>

<a id='scope'></a>
# B. Project Motivation
## Context
The goal: wrangle WeRateDogs Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations.

## The Data
### Enhanced Twitter Archive

The WeRateDogs Twitter archive contains basic tweet data for all 5000+ of their tweets, but not everything. One column the archive does contain though: each tweet's text, which I used to extract rating, dog name, and dog "stage" (i.e. doggo, floofer, pupper, and puppo) to make this Twitter archive "enhanced." Of the 5000+ tweets, I have filtered for tweets with ratings only (there are 2356).

![table of tweets analyzed](https://video.udacity-data.com/topher/2017/October/59dd4791_screenshot-2017-10-10-18.19.36/screenshot-2017-10-10-18.19.36.png)
*The extracted data from each tweet's text*

### Extracted data from tweet text
The extracted data from each tweet's text

This provided data set was extracted programmatically, but more processing (e.g., cleaning and tyding) is requried. The ratings probably aren't all correct. Same goes for the dog names and probably dog stages (see below for more information on these) too. As stated before more data processing is required to assess and clean these columns for later analysis and visualization.

![dog dictionary](https://video.udacity-data.com/topher/2017/October/59e04ceb_dogtionary-combined/dogtionary-combined.png)
*The Dogtionary explains the various stages of dog: doggo, pupper, puppo, and floof(er) (via the [#WeRateDogs book on Amazon](https://www.amazon.com/WeRateDogs-Most-Hilarious-Adorable-Youve/dp/1510717145))*

### Additional Data via the Twitter API

Back to the basic-ness of Twitter archives: retweet count and favorite count are two of the notable column omissions. Fortunately, this additional data can be gathered by anyone from Twitter's API. Well, "anyone" who has access to data for the 3000 most recent tweets, at least. The WeRateDogs Twitter archive and specifically the tweet IDs within it, can gather this data for all 5000+. The Twitter's API was used to query this valuable data. 

**Please note that the Twitter API was NOT utilized for this project for data securty/privacy reasons. This data was provided for the scope of this project.**

### Image Predictions File

One more cool thing: Every image in the WeRateDogs Twitter archive was processed through a [neural network](https://www.youtube.com/watch?v=2-Ol7ZB0MmU) that can classify breeds of dogs* (provided by project). The results: a table full of image predictions (the top three only) alongside each tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images).

![tweet image prediction](https://video.udacity-data.com/topher/2017/October/59dd4d2c_screenshot-2017-10-10-18.43.41/screenshot-2017-10-10-18.43.41.png)
*Tweet image prediction data*

### Image predictions
Tweet image prediction data

So for the last row in that table:

tweet_id is the last part of the tweet URL after "status/" → https://twitter.com/dog_rates/status/889531135344209921
* p1 is the algorithm's #1 prediction for the image in the tweet → **golden retriever**
* p1_conf is how confident the algorithm is in its #1 prediction → **95%**
* p1_dog is whether or not the #1 prediction is a breed of dog → **TRUE**
* p2 is the algorithm's second most likely prediction → **Labrador retriever**
* p2_conf is how confident the algorithm is in its #2 prediction → **1%**
* p2_dog is whether or not the #2 prediction is a breed of dog → **TRUE**
* etc.

And the #1 prediction for the image in that tweet was spot on:

@dog_rates tweet
![gold retriever](https://video.udacity-data.com/topher/2017/October/59dd4e05_dog-pred/dog-pred.png)
*A golden retriever named Stuart*

So that's all fun and good. But all of this additional data will need to be gathered, assessed, and cleaned--which is the scope of this project

## Key Points
Key points to keep in mind when data wrangling for this project:

* Only use original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html).
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* It is not required to gather tweets beyond August 1st, 2017 because it is out of scope. Image predictions cannot be gathered for new tweet data after this date because the source file for the image prediction is not provided--again out of scope for this project

<a href="#contents">[Table of Contents]</a>

<hr size="5"/>

<a id='gather'></a>
# 1. GATHERING DATA

<font color=blue>

## Gathering Data - Set 1 Requirements
**1.1 CRITERIA:** The student is able to gather data from a variety of sources and file formats.

**1.1 SPECIFICATION:**
Data is successfully gathered:
* From at least the three (3) different sources on the Project Details page.
* In at least the three (3) different file formats on the Project Details page.

Each piece of data is imported into a separate pandas DataFrame at first.

<a href="#contents">[Table of Contents]</a>

In [1]:
# import statements for all of the packages used for analysis

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
import statsmodels.api as sm;

# package required to get images from twitter data
import requests

# package required to access json file
import json


# Remember to include a 'magic word' so that your visualizations are plotted
#   inline with the notebook. See this page for more:
#   http://ipython.readthedocs.io/en/stable/interactive/magics.html

%matplotlib inline

### Gather Specification 1 - data provided (e.g., via flashdrive)

<a href="#gather">[Gathering Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

In [2]:
# Gather Specification 1 - file given (file provided via flashdrive)
df_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

### Gather Specification 2 - data accessed by internet via Requests library

<a href="#gather">[Gathering Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

In [3]:
# Gather Specification 2 - access via internet to get images for predictions, accessed programmatically
# https://pypi.org/project/requests/

# note that running this code will generate the file "image_prediction.tsv"

r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv',
                auth=('user','pass'))

# better check status and 
assert r.status_code == 200, 'Request corruption detected, status code is supposed to be 200 (aka "HTTP OK Success")'
assert r.encoding == 'utf-8', 'Request corruption detected, encoding was NOT \'utf-8\' (will cause issues later in json coding)'

open('image_prediction.tsv', 'wb').write(r.content)
df_prediction = pd.read_csv('image_prediction.tsv', sep='\t')
df_prediction.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

### Gather Specification 3 - data accessed by internet via Twitter API Tweepy

<a href="#gather">[Gathering Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

In [4]:
# Gather Specification 3 - json file
# https://stackoverflow.com/questions/47612822/how-to-create-pandas-dataframe-from-twitter-search-api
twitter_list =[]

with open('tweet-json.txt', encoding='utf-8') as json_file:
    for each_dictionary in json_file:   
        
        tweets_dict = {} # the dictionary terms are stored here
        tweets_json = json.loads(each_dictionary)
        
        tweets_dict['tweet_id'] = tweets_json['id_str']
        tweets_dict['retweet_count'] = tweets_json['retweet_count']
        tweets_dict['favorite_count'] = tweets_json['favorite_count']
        tweets_dict['full_text'] = tweets_json['full_text']
        
        twitter_list.append(tweets_dict)

# write to dataframe
df_twitter = pd.DataFrame(twitter_list)
df_twitter.head(3)

tweet_id  retweet_count  favorite_count  \
0  892420643555336193           8853           39467   
1  892177421306343426           6514           33819   
2  891815181378084864           4328           25461   

                                           full_text  
0  This is Phineas. He's a mystical boy. Only eve...  
1  This is Tilly. She's just checking pup on you....  
2  This is Archie. He is a rare Norwegian Pouncin...

<font color='red'>

**_Code below is required to programically interface and gather data from TWITTER using the ```Tweepy``` API. However, I used the provided file instead because I do NOT want to create Twitter account for cyber securty/privacy reasons._**

<font color='blue'>
             
> Essentially, you will run the Twitter API with will get you a JSON file (code block below). This JSON file will then be read using Python to get the ```tweet-json.txt``` used in the code block above. 


```
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = df_1.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)
```


## End of Gathering Data Section

<a href="#gather">[Gathering Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

<hr size="5"/>

<a id='assess'></a>
# 2. ASSESSING DATA

<font color=blue>

**2.1 CRITERIA:** The student is able to assess data visually and programmatically for quality and tidiness.

**2.1 SPECIFICATION:**
Two types of assessment are used:

* Visual assessment: each piece of gathered data is displayed in the Jupyter Notebook for visual assessment purposes. Once displayed, data can additionally be assessed in an external application (e.g. Excel, text editor).
* Programmatic assessment: pandas' functions and/or methods are used to assess the data.

---

**2.2 CRITERIA:** The student is able to thoroughly assess a dataset.

**2.2 SPECIFICATION:**
At least eight (8) data quality issues and two (2) tidiness issues are detected, and include the issues to clean to satisfy the Project Motivation. Each issue is documented in one to a few sentences each.

<a href="#contents">[Table of Contents]</a>

## Assess Specification 2.1 - Assessing data visual assessment (e.g., via CSV)

* Visual assessment: each piece of gathered data is displayed in the Jupyter Notebook for visual assessment purposes. Once displayed, data can additionally be assessed in an external application (e.g. Excel, text editor).

<a href="#assess">[Assessing Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

Current dataframes active from **_Section 1. GATHER_**
1. ```df_archive``` = tweet metadata
2. ```df_prediction``` = tweet picture prediction
3. ```df_twitter``` = tweet retweet_count, favorite_count, and full_text

In [5]:
# visual assessment - warning to user that opening these files "may" cause viewing software to crash
# help: https://www.guru99.com/python-check-if-file-exists.html

import os.path
from os import path
fileFlag = False # initilized as FALSE -- assuming .csv files have not be created from dataframes in this notebook

In [6]:
# help: https://www.geeksforgeeks.org/g-fact-41-multiple-return-values-in-python/
def createCSV():
    # adding "1-" for easier file handling in folder
    df_archive.to_csv('1-archive.csv')
    df_prediction.to_csv('1-prediction.csv')
    df_twitter.to_csv('1-twitter.csv')
    fileFlag = True
    print('.csv Files created ... Ready for user VISUAL assessment')
    return fileFlag

if path.exists('1-archive.csv') == True and path.exists('1-prediction.csv') == True and path.exists('1-twitter.csv') == True:
    print("Files aready created! ... Ready for user VISUAL assessment")
elif fileFlag == False:
    fileFlag = createCSV() # Assign returned tuple , execute file create assuming fileFlag is FALSE (has not be done)
else:
    assert path.exists('1-archive.csv') == True, "You need to create the 1-archive.csv file"
    assert path.exists('1-prediction.csv') == True, "You need to create the 1-prediction.csv file"
    assert path.exists('1-twitter.csv') == True, "You need to create the 1-twitter.csv file"
    print("Files exists! ... Ready for user VISUAL assessment")

# could add button here for "confirmation" data was reviewed

Files aready created! ... Ready for user VISUAL assessment


## Assess Specification 2.1 - Assessing data PROGRAMMICALLY assessment (e.g., via CSV)

* Programmatic assessment: pandas' functions and/or methods are used to assess the data.

<a href="#assess">[Assessing Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

In [7]:
# Programmically assessment of data

#import the newly created .csv files
df_archive = pd.read_csv('1-archive.csv')
df_prediction = pd.read_csv('1-prediction.csv')
df_twitter = pd.read_csv('1-twitter.csv')

<font color='blue'>

Review ```df_archive``` data programmatically

In [8]:
df_archive.head(3)

Unnamed: 0            tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0           0  892420643555336193                    NaN                  NaN   
1           1  892177421306343426                    NaN                  NaN   
2           2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [9]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 18 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  2356 non-null   int64  
 1   tweet_id                    2356 non-null   int64  
 2   in_reply_to_status_id       78 non-null     float64
 3   in_reply_to_user_id         78 non-null     float64
 4   timestamp                   2356 non-null   object 
 5   source                      2356 non-null   object 
 6   text                        2356 non-null   object 
 7   retweeted_status_id         181 non-null    float64
 8   retweeted_status_user_id    181 non-null    float64
 9   retweeted_status_timestamp  181 non-null    object 
 10  expanded_urls               2297 non-null   object 
 11  rating_numerator            2356 non-null   int64  
 12  rating_denominator          2356 non-null   int64  
 13  name                        2356 

<font color='blue'>

Review ```df_prediction``` data programmatically

In [10]:
df_prediction.head(3)

Unnamed: 0            tweet_id  \
0           0  666020888022790149   
1           1  666029285002620928   
2           2  666033412701032449   

                                           jpg_url  img_num  \
0  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg        1   
1  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg        1   
2  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg        1   

                       p1   p1_conf  p1_dog                  p2   p2_conf  \
0  Welsh_springer_spaniel  0.465074    True              collie  0.156665   
1                 redbone  0.506826    True  miniature_pinscher  0.074192   
2         German_shepherd  0.596461    True            malinois  0.138584   

   p2_dog                   p3   p3_conf  p3_dog  
0    True    Shetland_sheepdog  0.061428    True  
1    True  Rhodesian_ridgeback  0.072010    True  
2    True           bloodhound  0.116197    True

In [11]:
df_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  2075 non-null   int64  
 1   tweet_id    2075 non-null   int64  
 2   jpg_url     2075 non-null   object 
 3   img_num     2075 non-null   int64  
 4   p1          2075 non-null   object 
 5   p1_conf     2075 non-null   float64
 6   p1_dog      2075 non-null   bool   
 7   p2          2075 non-null   object 
 8   p2_conf     2075 non-null   float64
 9   p2_dog      2075 non-null   bool   
 10  p3          2075 non-null   object 
 11  p3_conf     2075 non-null   float64
 12  p3_dog      2075 non-null   bool   
dtypes: bool(3), float64(3), int64(3), object(4)
memory usage: 168.3+ KB


<font color='blue'>

Review ```df_twitter``` data programmatically

In [12]:
df_twitter.head(3)

Unnamed: 0            tweet_id  retweet_count  favorite_count  \
0           0  892420643555336193           8853           39467   
1           1  892177421306343426           6514           33819   
2           2  891815181378084864           4328           25461   

                                           full_text  
0  This is Phineas. He's a mystical boy. Only eve...  
1  This is Tilly. She's just checking pup on you....  
2  This is Archie. He is a rare Norwegian Pouncin...

In [13]:
df_twitter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Unnamed: 0      2354 non-null   int64 
 1   tweet_id        2354 non-null   int64 
 2   retweet_count   2354 non-null   int64 
 3   favorite_count  2354 non-null   int64 
 4   full_text       2354 non-null   object
dtypes: int64(4), object(1)
memory usage: 92.1+ KB


## Assess Specification 2.2 - Identify data quality and tidiness issues and steps for cleaning

* At least eight (8) data quality issues and two (2) tidiness issues are detected, and include the issues to clean to satisfy the Project Motivation. Each issue is documented in one to a few sentences each.

**Tips for Tidying**
1. Each variable forms a column.
2. Each observation forms a row.
3. Each type of observational unit forms a table
*Reference for [tidy data here](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html)*

**Tips for Common Data Quality Issues**
1. Missing data
2. Invalide data (e.g., state a negative height, or other datatype validation errors--str vs int vs float, think there can only be 2 people in a room... not 2.54 people in a room... *unless there's ghosts lol*)
3. Inaccurate data (e.g., specifying a foot = 5 inches, which is WRONG. A foot = 12 inches)
4. Inconsistent data (e.g., mixing up units, some data captured as cm instead of inches)

<a href="#assess">[Assessing Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

In [14]:
# another look at the data, determine what cleaning steps are required
print('archive data\n')
df_archive.info()
print('______________________________________________________________\n')

print('prediction data\n')
df_prediction.info()
print('______________________________________________________________\n')

print('twitter data\n')
df_twitter.info()
print('______________________________________________________________\n')


archive data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 18 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  2356 non-null   int64  
 1   tweet_id                    2356 non-null   int64  
 2   in_reply_to_status_id       78 non-null     float64
 3   in_reply_to_user_id         78 non-null     float64
 4   timestamp                   2356 non-null   object 
 5   source                      2356 non-null   object 
 6   text                        2356 non-null   object 
 7   retweeted_status_id         181 non-null    float64
 8   retweeted_status_user_id    181 non-null    float64
 9   retweeted_status_timestamp  181 non-null    object 
 10  expanded_urls               2297 non-null   object 
 11  rating_numerator            2356 non-null   int64  
 12  rating_denominator          2356 non-null   int64  
 13  name               

In [15]:
print('Any duplicated records in df_archive? -->', df_archive.duplicated().sum())
print('Any duplicated records in df_prediction? --> ', df_prediction.duplicated().sum())
print('Any duplicated records in df_twitter? --> ',df_twitter.duplicated().sum())

Any duplicated records in df_archive? --> 0
Any duplicated records in df_prediction? -->  0
Any duplicated records in df_twitter? -->  0


In [16]:
print('Any missing records in df_archive?\n', df_archive.isnull().sum())
print('\nAny missing records in df_prediction?\n', df_prediction.isnull().sum())
print('\nAny missing records in df_twitter?\n',df_twitter.isnull().sum())

Any missing records in df_archive?
 Unnamed: 0                       0
tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

Any missing records in df_prediction?
 Unnamed: 0    0
tweet_id      0
jpg_url       0
img_num       0
p1            0
p1_conf       0
p1_dog        0
p2            0
p2_conf       0
p2_dog        0
p3            0
p3_conf       0
p3_dog        0
dtype: int64

Any missing records in df_twitter?
 Unnamed: 0        0
twe

In [17]:
print('archive data\n')
df_archive.nunique()

archive data



Unnamed: 0                    2356
tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [18]:
print('prediction data\n')
df_prediction.nunique()

prediction data



Unnamed: 0    2075
tweet_id      2075
jpg_url       2009
img_num          4
p1             378
p1_conf       2006
p1_dog           2
p2             405
p2_conf       2004
p2_dog           2
p3             408
p3_conf       2006
p3_dog           2
dtype: int64

In [19]:
print('twitter data\n')
df_twitter.nunique()

twitter data



Unnamed: 0        2354
tweet_id          2354
retweet_count     1724
favorite_count    2007
full_text         2354
dtype: int64

In [20]:
# quick looks at statis to determine spread to eliminate outliers
ratings_level = 100*(df_archive.rating_numerator / df_archive.rating_denominator)
ratings_level
ratings_level.describe()

# so, lets make the cut-off point for exceedingly high levels of ratings be 120%... 
# but that deletes over 400 records... so lets say 150% then! Only 12 records removed!

count    2356.0
mean        inf
std         NaN
min         0.0
25%       100.0
50%       110.0
75%       120.0
max         inf
dtype: float64

In [21]:
# quick look at the stats of retweets and favorites to understand spread to remove "zero" values
df_twitter.describe()

# actually, lets not remove anything, this is data from Twitter database, 
# these "zero" because low probability that they were mistakes 

Unnamed: 0      tweet_id  retweet_count  favorite_count
count  2354.000000  2.354000e+03    2354.000000     2354.000000
mean   1176.500000  7.426978e+17    3164.797366     8080.968564
std     679.685589  6.852812e+16    5284.770364    11814.771334
min       0.000000  6.660209e+17       0.000000        0.000000
25%     588.250000  6.783975e+17     624.500000     1415.000000
50%    1176.500000  7.194596e+17    1473.500000     3603.500000
75%    1764.750000  7.993058e+17    3652.000000    10122.250000
max    2353.000000  8.924206e+17   79515.000000   132810.000000

<font color='red'>

<a id='todo'></a>
## QUALITY (click hyperlink question to go to section in notebook!)

<a href="#quality-section">[Go to QUALITY section]</a>
    
1. <a href="#Q1">[all dataframes --> Remove missing data column ```Unnamed: 0```]</a>
    
    
2. <a href="#Q2">[Missing data, Drop columns:]</a>
    * df_archive
        * in_reply_to_status_id
        * in_reply_to_user_id
        * retweeted_status_id
        * retweeted_status_user_id
        * retweeted_status_timestamp
        * expanded_urls 
        * in_reply_to_status_id
        * in_reply_to_user_id 
    
    
3. <a href="#Q3">[df_archive --> convert columns to boolean for easier analysis]</a> 
    * doggo
    * floofer
    * pupper
    * puppo

    
4. <a href="#Q4">[df_archive --> convert form object to in64]</a>
    * doggo
    * floofer
    * pupper
    * puppo
    
    
5. <a href="#Q5">[df_archive --> need for classification of ```floof``` for dogs that did not meeting the four listed catagories below:]</a>
    * doggo
    * floofer
    * pupper
    * puppo
    
    
6. <a href="#Q6">[df_archive --> remove entries with EXTREME ratings outliers (anything above 150%). Those entries are:]</a> 
    * 55 with 170%
    * 188 with 4200%
    * 189 with 6660%
    * 290 with 1820%
    * 340 with 750%
    * 516 with 343%
    * 695 with 750%
    * 763 with 270%
    * 979 with 17760% wow!
    * 1712 with 260%
    * 2074 with 4200%

    
7. <a href="#Q7">[df_archive --> remove entry 313 because no denominator is provided (entered as 0)]</a>

    
8. <a href="#Q8">[df_prediction --> make all names of dogs CAPITALIZED, to prevent any variants of different capitilizations]</a>

>*At least eight (8) data quality issues and two (2) tidiness issues are detected, and include the issues to clean to satisfy the Project Motivation. Each issue is documented in one to a few sentences each.*
    
>>**Tips for Common Data Quality Issues**
>>1. Missing data
2. Invalide data (e.g., state a negative height, or other datatype validation errors--str vs int vs float, think there can only be 2 people in a room... not 2.54 people in a room... *unless there's ghosts lol*)
3. Inaccurate data (e.g., specifying a foot = 5 inches, which is WRONG. A foot = 12 inches)
4. Inconsistent data (e.g., mixing up units, some data captured as cm instead of inches)



## TIDINESS
    
<a href="#tidy-section">[Go to TIDY section]</a>
    
1. <a href="#T1">[Combine all seperated tables, join by ```tweet_id```]</a>

2. <a href="#T2">[df_twitter drop column ```full_text``` because it contains multiple observational units (e.g., name, url, score, etc.)]</a>

>*At least eight (8) data quality issues and two (2) tidiness issues are detected, and include the issues to clean to satisfy the Project Motivation. Each issue is documented in one to a few sentences each.*
    
>>**Tips for Tidying**
>>1. Each variable forms a column.
2. Each observation forms a row.
3. Each type of observational unit forms a table
*Reference for [tidy data here](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html)*
    
## FEATURE ENGINEEERING
1. df_archive
    * create a percentage column of the numerator and denometer ratings
    * create column for twitter-web vs twitter-phone vs twitter-tweetdeck vs vine source
    * create a column for "bonus" for any ratings that exceed score of 10
    
<a href="#assess">[Assessing Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

<hr size="5"/>

<a id='clean'></a>
# 3. CLEANING DATA

<font color=blue>
    
**3.1 CRITERIA:** The student uses the steps in the data cleaning process to guide their cleaning efforts.

**3.1 SPECIFICATION:**
The define, code, and test steps of the cleaning process are clearly documented.

---
    
**3.2 CRITERIA:** The student is able to thoroughly clean a dataset programmatically.

**3.2 SPECIFICATION:**

Copies of the original pieces of data are made prior to cleaning.

All issues identified in the assess phase are successfully cleaned (if possible) using Python and pandas, and include the cleaning tasks required to satisfy the Project Motivation.

A tidy master dataset (or datasets, if appropriate) with all pieces of gathered data is created.
    
<a href="#todo">[Cleaning and Tidying To-do List]</a> <a href="#contents">[Table of Contents]</a>

## Cleaning Specification 3.1 and 3.2 - Cleaning Data (notes captured in codeblocks and markdown as required)

<a href="#clean">[Cleaning Data Requirements]</a> 
<a href="#todo">[Cleaning and Tidying To-do List]</a> 
<a href="#contents">[Table of Contents]</a>

<a id='quality-section'></a>

<font color='red'>
    
## QUALITY ISSUES ADDRESSED -- note that ```df``` will change to ```dfc``` to indicate cleaned data
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

<a id='Q1'></a>

<font color='red'>

### ✔️ 1. all dataframes --> Remove missing data column ```Unnamed: 0```

<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [22]:
# before cleaning
print("df_archive data column names\n", list(df_archive.columns))
print("\ndf_prediciton data column names\n", list(df_prediction.columns))
print("\ndf_twitter data column names\n", list(df_twitter.columns))

df_archive data column names
 ['Unnamed: 0', 'tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo']

df_prediciton data column names
 ['Unnamed: 0', 'tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog']

df_twitter data column names
 ['Unnamed: 0', 'tweet_id', 'retweet_count', 'favorite_count', 'full_text']


In [23]:
# all data frames drop column "Unnamed: 0", assign new df names -- incase we screw something up
dfc_archive = df_archive.drop(columns = ['Unnamed: 0']);
dfc_prediction = df_prediction.drop(columns = ['Unnamed: 0']);
dfc_twitter = df_twitter.drop(columns = ['Unnamed: 0']);

In [24]:
# verification that headers removed
print("dfc_archive data column names\n", list(dfc_archive.columns))
print("\ndfc_prediciton data column names\n", list(dfc_prediction.columns))
print("\ndfc_twitter data column names\n", list(dfc_twitter.columns))

dfc_archive data column names
 ['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo']

dfc_prediciton data column names
 ['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog']

dfc_twitter data column names
 ['tweet_id', 'retweet_count', 'favorite_count', 'full_text']


<a id='Q2'></a>

<font color='red'>
    
### ✔️ 2. Missing data, Drop columns:

* df_archive
    * in_reply_to_status_id
    * in_reply_to_user_id
    * retweeted_status_id
    * retweeted_status_user_id
    * retweeted_status_timestamp
    * expanded_urls 
    * in_reply_to_status_id
    * in_reply_to_user_id 
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [25]:
# before cleaning
print("dfc_archive data column names\n", list(dfc_archive.columns))

dfc_archive data column names
 ['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo']


In [26]:
# drop columns with missing data because they do not provide value at this point
dfc_archive = dfc_archive.drop(columns = ['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id',
                                         'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls',
                                         'in_reply_to_status_id', 'in_reply_to_user_id']);

In [27]:
# verification after cleaning
print("dfc_archive data column names\n", list(dfc_archive.columns))

dfc_archive data column names
 ['tweet_id', 'timestamp', 'source', 'text', 'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo']


<a id='Q3'></a>

<font color='red'>

    
### ✔️  3. df_archive --> convert columns to boolean for easier analysis
* doggo
* floofer
* pupper
* puppo
   
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [28]:
# visual review dfc_archive
dfc_archive.head(3)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [29]:
# programmically review the counts for these dog columns
print('DOGGO column counts\n', dfc_archive.doggo.value_counts())
print('____________')
print('\nFLOOFER column counts\n', dfc_archive.floofer.value_counts())
print('____________')
print('\nPUPPER column counts\n', dfc_archive.pupper.value_counts())
print('____________')
print('\nPUPPO column counts\n', dfc_archive.puppo.value_counts())

DOGGO column counts
 None     2259
doggo      97
Name: doggo, dtype: int64
____________

FLOOFER column counts
 None       2346
floofer      10
Name: floofer, dtype: int64
____________

PUPPER column counts
 None      2099
pupper     257
Name: pupper, dtype: int64
____________

PUPPO column counts
 None     2326
puppo      30
Name: puppo, dtype: int64


In [30]:
# check datatype
dfc_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2356 non-null   int64 
 1   timestamp           2356 non-null   object
 2   source              2356 non-null   object
 3   text                2356 non-null   object
 4   rating_numerator    2356 non-null   int64 
 5   rating_denominator  2356 non-null   int64 
 6   name                2356 non-null   object
 7   doggo               2356 non-null   object
 8   floofer             2356 non-null   object
 9   pupper              2356 non-null   object
 10  puppo               2356 non-null   object
dtypes: int64(3), object(8)
memory usage: 202.6+ KB


In [31]:
# change dfc_archive to test for prototyping solution
test = dfc_archive

In [32]:
# replace all "None" with 0
# help: https://www.edureka.co/community/584/how-can-replace-values-with-none-in-a-dataframe-using-pandas
df1 = test.replace('None', 0)

df1.head(3)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas     0       0      0     0  
1                  10    Tilly     0       0      0     0  
2                  10   Archie     0       0      0     0

In [33]:
print('"DOG name" column name count with "None" replaced with zero\n')
print('DOGGO column counts\n', df1.doggo.value_counts())
print('____________')
print('\nFLOOFER column counts\n', df1.floofer.value_counts())
print('____________')
print('\nPUPPER column counts\n', df1.pupper.value_counts())
print('____________')
print('\nPUPPO column counts\n', df1.puppo.value_counts())

"DOG name" column name count with "None" replaced with zero

DOGGO column counts
 0        2259
doggo      97
Name: doggo, dtype: int64
____________

FLOOFER column counts
 0          2346
floofer      10
Name: floofer, dtype: int64
____________

PUPPER column counts
 0         2099
pupper     257
Name: pupper, dtype: int64
____________

PUPPO column counts
 0        2326
puppo      30
Name: puppo, dtype: int64


In [34]:
# replace all "dog column name" with 1, could use function here...
df1 = df1.replace('doggo', 1)
df1 = df1.replace('floofer', 1)
df1 = df1.replace('pupper', 1)
df1 = df1.replace('puppo', 1)

In [35]:
print('"DOG name" column name count with "DOG name" replaced with one\n')
print('DOGGO column counts\n', df1.doggo.value_counts())
print('____________')
print('\nFLOOFER column counts\n', df1.floofer.value_counts())
print('____________')
print('\nPUPPER column counts\n', df1.pupper.value_counts())
print('____________')
print('\nPUPPO column counts\n', df1.puppo.value_counts())

"DOG name" column name count with "DOG name" replaced with one

DOGGO column counts
 0    2259
1      97
Name: doggo, dtype: int64
____________

FLOOFER column counts
 0    2346
1      10
Name: floofer, dtype: int64
____________

PUPPER column counts
 0    2099
1     257
Name: pupper, dtype: int64
____________

PUPPO column counts
 0    2326
1      30
Name: puppo, dtype: int64


<a id='Q4'></a>

<font color='red'>

    
### ✔️ 4. df_archive --> convert form object to in64

* doggo
* floofer
* pupper
* puppo
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [36]:
# method executed above converted all the datatypes to integers!
# but... the code below will convert those "DOG name" columns to 
# integers, lets also change the df5 name back to dfc_archive
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2356 non-null   int64 
 1   timestamp           2356 non-null   object
 2   source              2356 non-null   object
 3   text                2356 non-null   object
 4   rating_numerator    2356 non-null   int64 
 5   rating_denominator  2356 non-null   int64 
 6   name                2356 non-null   object
 7   doggo               2356 non-null   int64 
 8   floofer             2356 non-null   int64 
 9   pupper              2356 non-null   int64 
 10  puppo               2356 non-null   int64 
dtypes: int64(7), object(4)
memory usage: 202.6+ KB


In [37]:
# stirng to integer datatype
df1['doggo'] = df1['doggo'].astype(int)
df1['floofer'] = df1['floofer'].astype(int)
df1['pupper'] = df1['pupper'].astype(int)
df1['puppo'] = df1['puppo'].astype(int)

In [38]:
dfc_archive = df1
dfc_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2356 non-null   int64 
 1   timestamp           2356 non-null   object
 2   source              2356 non-null   object
 3   text                2356 non-null   object
 4   rating_numerator    2356 non-null   int64 
 5   rating_denominator  2356 non-null   int64 
 6   name                2356 non-null   object
 7   doggo               2356 non-null   int32 
 8   floofer             2356 non-null   int32 
 9   pupper              2356 non-null   int32 
 10  puppo               2356 non-null   int32 
dtypes: int32(4), int64(3), object(4)
memory usage: 165.8+ KB


In [39]:
# quick visual check --> looks good!
dfc_archive.sample(5)

tweet_id                  timestamp  \
1078  739485634323156992  2016-06-05 15:54:48 +0000   
304   836380477523124226  2017-02-28 01:00:19 +0000   
105   872122724285648897  2017-06-06 16:07:15 +0000   
794   773336787167145985  2016-09-07 01:47:12 +0000   
342   832088576586297345  2017-02-16 04:45:50 +0000   

                                                 source  \
1078  <a href="http://twitter.com/download/iphone" r...   
304   <a href="http://twitter.com/download/iphone" r...   
105   <a href="http://twitter.com/download/iphone" r...   
794   <a href="http://twitter.com/download/iphone" r...   
342   <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1078  This is Kyle. He's a heavy drinker and an avid...                 6   
304   This is Ava. She just blasted off. Streamline ...                12   
105   This is Kody. He's a baller. Wishes he was a l...                12   
794   RT @dog_rates: Meet Fizz. She thinks love is a...                11   
342            @docmisterio account started on 11/15/15                11   

      rating_denominator  name  doggo  floofer  pupper  puppo  
1078                  10  Kyle      0        0       0      0  
304                   10   Ava      0        0       1      0  
105                   10  Kody      0        0       0      0  
794                   10  Fizz      0        0       0      0  
342                   15     0      0        0       0      0

---

<a id='Q5'></a>

<font color='red'>

    
### ✔️ 5. df_archive --> need for classification of ```floof``` for dogs that did not meeting the four listed catagories below:
        
* doggo
* floofer
* pupper
* puppo
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [40]:
dfc_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2356 non-null   int64 
 1   timestamp           2356 non-null   object
 2   source              2356 non-null   object
 3   text                2356 non-null   object
 4   rating_numerator    2356 non-null   int64 
 5   rating_denominator  2356 non-null   int64 
 6   name                2356 non-null   object
 7   doggo               2356 non-null   int32 
 8   floofer             2356 non-null   int32 
 9   pupper              2356 non-null   int32 
 10  puppo               2356 non-null   int32 
dtypes: int32(4), int64(3), object(4)
memory usage: 165.8+ KB


In [41]:
# creating test dataframe for prototyping
test = dfc_archive

In [42]:
# help: https://stackoverflow.com/questions/26886653/pandas-create-new-column-based-on-values-from-other-columns-apply-a-function-o
def Dog_Class (row):
   if row['doggo'] == 1 :
      return 'doggo'
   if row['floofer'] == 1 :
      return 'floofer'
   if row['pupper'] == 1:
      return 'pupper'
   if row['puppo']  == 1:
      return 'puppo'
   if row['doggo'] + row['floofer'] + row['pupper'] + row['puppo'] > 1 :
      return 'Two or more? What the woof!? Going to find you a home!'
   return 'floof'

In [43]:
test.apply (lambda row: Dog_Class(row), axis=1)

0       floof
1       floof
2       floof
3       floof
4       floof
        ...  
2351    floof
2352    floof
2353    floof
2354    floof
2355    floof
Length: 2356, dtype: object

In [44]:
# apply new series as column in dataframe
test['Dog_Class'] = test.apply(lambda row: Dog_Class(row), axis=1)
test.sample(5)

tweet_id                  timestamp  \
1571  687704180304273409  2016-01-14 18:33:48 +0000   
1199  716791146589110272  2016-04-04 00:55:01 +0000   
1106  734787690684657664  2016-05-23 16:46:51 +0000   
59    880872448815771648  2017-06-30 19:35:32 +0000   
312   835264098648616962  2017-02-24 23:04:14 +0000   

                                                 source  \
1571  <a href="http://twitter.com/download/iphone" r...   
1199  <a href="http://twitter.com/download/iphone" r...   
1106  <a href="http://twitter.com/download/iphone" r...   
59    <a href="http://twitter.com/download/iphone" r...   
312   <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1571  Say hello to Blakely. He thinks that's a hat. ...                 9   
1199  Meet Jennifur. She's supposed to be navigating...                11   
1106  This dog is more successful than I will ever b...                13   
59    Ugh not again. We only rate dogs. Please don't...                12   
312   Meet Lola. Her hobbies include being precious ...                12   

      rating_denominator      name  doggo  floofer  pupper  puppo Dog_Class  
1571                  10   Blakely      0        0       1      0    pupper  
1199                  10  Jennifur      0        0       0      0     floof  
1106                  10         0      0        0       0      0     floof  
59                    10         0      0        0       0      0     floof  
312                   10      Lola      0        0       0      0     floof

In [45]:
# assign test to dfc_archive again
dfc_archive = test
dfc_archive.head(2)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   

   rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  
0                  10  Phineas      0        0       0      0     floof  
1                  10    Tilly      0        0       0      0     floof

---

<a id='Q6'></a>

<font color='red'>


### ✔️ 6. df_archive --> remove entries with EXTREME ratings outliers (anything above 150%). Those entries are (checking sheet): 
             
* 55 with 170%
* 188 with 4200%
* 189 with 6660%
* 290 with 1820%
* 340 with 750%
* 516 with 343%
* 695 with 750%
* 763 with 270%
* 979 with 17760% wow!
* 1712 with 260%
* 2074 with 4200%
    
But we are going to do this programmically! ;)
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [46]:
# create column for ratings percentage
test = dfc_archive
test['ratings_percentage'] = test['rating_numerator'] / test['rating_denominator']
test.sample(3)

tweet_id                  timestamp  \
1385  700747788515020802  2016-02-19 18:24:26 +0000   
1009  747512671126323200  2016-06-27 19:31:23 +0000   
2345  666063827256086533  2015-11-16 01:22:45 +0000   

                                                 source  \
1385  <a href="http://twitter.com/download/iphone" r...   
1009  <a href="http://twitter.com/download/iphone" r...   
2345  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1385  We only rate dogs. Pls stop sending in non-can...                11   
1009  Guys pls stop sending actual sharks. It's too ...                10   
2345  This is the happiest dog you will ever see. Ve...                10   

      rating_denominator  name  doggo  floofer  pupper  puppo Dog_Class  \
1385                  10  very      0        0       0      0     floof   
1009                  10     0      0        0       0      0     floof   
2345                  10   the      0        0       0      0     floof   

      ratings_percentage  
1385                 1.1  
1009                 1.0  
2345                 1.0

In [47]:
# these rows we know need to go--we will verify later...
# help: https://www.shanelynn.ie/select-pandas-dataframe-rows-and-columns-using-iloc-loc-and-ix/

drop = dfc_archive.iloc[[55, 188, 189, 290, 340, 516, 695, 763, 979, 1712, 2074],:]
drop

tweet_id                  timestamp  \
55    881633300179243008  2017-07-02 21:58:53 +0000   
188   855862651834028034  2017-04-22 19:15:32 +0000   
189   855860136149123072  2017-04-22 19:05:32 +0000   
290   838150277551247360  2017-03-04 22:12:52 +0000   
340   832215909146226688  2017-02-16 13:11:49 +0000   
516   810984652412424192  2016-12-19 23:06:23 +0000   
695   786709082849828864  2016-10-13 23:23:56 +0000   
763   778027034220126208  2016-09-20 00:24:34 +0000   
979   749981277374128128  2016-07-04 15:00:45 +0000   
1712  680494726643068929  2015-12-25 21:06:00 +0000   
2074  670842764863651840  2015-11-29 05:52:33 +0000   

                                                 source  \
55    <a href="http://twitter.com/download/iphone" r...   
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1712  <a href="http://twitter.com/download/iphone" r...   
2074  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
55    @roushfenway These are good dogs but 17/10 is ...                17   
188   @dhmontgomery We also gave snoop dogg a 420/10...               420   
189   @s8n You tried very hard to portray this good ...               666   
290                                  @markhoppus 182/10               182   
340   RT @dog_rates: This is Logan, the Chow who liv...                75   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                24   
695   This is Logan, the Chow who lived. He solemnly...                75   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                27   
979   This is Atticus. He's quite simply America af....              1776   
1712  Here we have uncovered an entire battalion of ...                26   
2074  After so many requests... here you go.\n\nGood...               420   

      rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  \
55                    10        0      0        0       0      0     floof   
188                   10        0      0        0       0      0     floof   
189                   10        0      0        0       0      0     floof   
290                   10        0      0        0       0      0     floof   
340                   10    Logan      0        0       0      0     floof   
516                    7      Sam      0        0       0      0     floof   
695                   10    Logan      0        0       0      0     floof   
763                   10   Sophie      0        0       1      0    pupper   
979                   10  Atticus      0        0       0      0     floof   
1712                  10        0      0        0       0      0     floof   
2074                  10        0      0        0       0      0     floof   

      ratings_percentage  
55              1.700000  
188            42.000000  
189            66.600000  
290            18.200000  
340             7.500000  
516             3.428571  
695             7.500000  
763             2.700000  
979           177.600000  
1712            2.600000  
2074           42.000000

In [48]:
print('shape of dataset', test.shape)
print('unique value count check for "Ratings_Percentage" = ', test.ratings_percentage.nunique())

shape of dataset (2356, 13)
unique value count check for "Ratings_Percentage" =  30


In [49]:
# BEFORE DROPPING ROWS
# status review -- there is an undefine value
test.describe() 

tweet_id  rating_numerator  rating_denominator        doggo  \
count  2.356000e+03       2356.000000         2356.000000  2356.000000   
mean   7.427716e+17         13.126486           10.455433     0.041171   
std    6.856705e+16         45.876648            6.745237     0.198729   
min    6.660209e+17          0.000000            0.000000     0.000000   
25%    6.783989e+17         10.000000           10.000000     0.000000   
50%    7.196279e+17         11.000000           10.000000     0.000000   
75%    7.993373e+17         12.000000           10.000000     0.000000   
max    8.924206e+17       1776.000000          170.000000     1.000000   

           floofer       pupper        puppo  ratings_percentage  
count  2356.000000  2356.000000  2356.000000              2356.0  
mean      0.004244     0.109083     0.012733                 inf  
std       0.065025     0.311810     0.112146                 NaN  
min       0.000000     0.000000     0.000000                 0.0  
25%       0.000000     0.000000     0.000000                 1.0  
50%       0.000000     0.000000     0.000000                 1.1  
75%       0.000000     0.000000     0.000000                 1.2  
max       1.000000     1.000000     1.000000                 inf

<a id='Q7_solved'></a>

<font color='red'>
Note that this method here solves Quality Cleaning #7

<a href="#Q7">[Quality Cleaning #7 question]</a>

In [50]:
# drop rows over 10k for now, to remove infinite value
test_filtered = test[test['ratings_percentage'] <= 100000] 

print('shape of dataset', test_filtered.shape)
print('unique value count check for "Ratings_Percentage" = ', test_filtered.ratings_percentage.nunique())

shape of dataset (2355, 13)
unique value count check for "Ratings_Percentage" =  29


In [51]:
test_filtered.describe()

tweet_id  rating_numerator  rating_denominator        doggo  \
count  2.355000e+03       2355.000000         2355.000000  2355.000000   
mean   7.427323e+17         12.724416           10.459873     0.041189   
std    6.855511e+16         41.527443            6.743226     0.198769   
min    6.660209e+17          0.000000            2.000000     0.000000   
25%    6.783982e+17         10.000000           10.000000     0.000000   
50%    7.195514e+17         11.000000           10.000000     0.000000   
75%    7.993029e+17         12.000000           10.000000     0.000000   
max    8.924206e+17       1776.000000          170.000000     1.000000   

           floofer       pupper        puppo  ratings_percentage  
count  2355.000000  2355.000000  2355.000000         2355.000000  
mean      0.004246     0.109130     0.012739            1.222032  
std       0.065039     0.311868     0.112169            4.083485  
min       0.000000     0.000000     0.000000            0.000000  
25%       0.000000     0.000000     0.000000            1.000000  
50%       0.000000     0.000000     0.000000            1.100000  
75%       0.000000     0.000000     0.000000            1.200000  
max       1.000000     1.000000     1.000000          177.600000

In [52]:
# wow, ok, we need to remove any ratings over 177,600%! But lets filter out anything greater than 150%... so 1.5
test_filtered1 = test_filtered[test_filtered['ratings_percentage'] <= 1.5] 

print('shape of dataset', test_filtered1.shape)
print('unique value count check for "Ratings_Percentage" = ', test_filtered1.ratings_percentage.nunique())

shape of dataset (2344, 13)
unique value count check for "Ratings_Percentage" =  20


In [53]:
test_filtered1.describe()

tweet_id  rating_numerator  rating_denominator        doggo  \
count  2.344000e+03       2344.000000         2344.000000  2344.000000   
mean   7.424888e+17         11.202218           10.463311     0.041382   
std    6.846984e+16          7.988365            6.758591     0.199215   
min    6.660209e+17          0.000000            2.000000     0.000000   
25%    6.783868e+17         10.000000           10.000000     0.000000   
50%    7.191522e+17         11.000000           10.000000     0.000000   
75%    7.989278e+17         12.000000           10.000000     0.000000   
max    8.924206e+17        204.000000          170.000000     1.000000   

           floofer       pupper        puppo  ratings_percentage  
count  2344.000000  2344.000000  2344.000000         2344.000000  
mean      0.004266     0.109215     0.012799            1.069137  
std       0.065191     0.311975     0.112429            0.219015  
min       0.000000     0.000000     0.000000            0.000000  
25%       0.000000     0.000000     0.000000            1.000000  
50%       0.000000     0.000000     0.000000            1.100000  
75%       0.000000     0.000000     0.000000            1.200000  
max       1.000000     1.000000     1.000000            1.500000

In [54]:
# verify that the following rows below are removed
# range check: [54:56, 187:190, 289:291, 339:341, 515:517, 694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [55, 188, 189, 290, 340, 516, 695, 763, 979, 1712, 2074]
test_filtered1.iloc[[55, 188, 189, 290, 340, 516, 695, 763, 979, 1712, 2074]]

tweet_id                  timestamp  \
56    881536004380872706  2017-07-02 15:32:16 +0000   
191   855851453814013952  2017-04-22 18:31:02 +0000   
192   855818117272018944  2017-04-22 16:18:34 +0000   
294   837482249356513284  2017-03-03 01:58:22 +0000   
346   831926988323639298  2017-02-15 18:03:45 +0000   
523   809448704142938112  2016-12-15 17:23:04 +0000   
703   785927819176054784  2016-10-11 19:39:28 +0000   
772   776477788987613185  2016-09-15 17:48:25 +0000   
989   748932637671223296  2016-07-01 17:33:49 +0000   
1723  680100725817409536  2015-12-24 19:00:23 +0000   
2086  670803562457407488  2015-11-29 03:16:46 +0000   

                                                 source  \
56    <a href="http://twitter.com/download/iphone" r...   
191   <a href="http://twitter.com/download/iphone" r...   
192   <a href="http://twitter.com/download/iphone" r...   
294   <a href="http://twitter.com/download/iphone" r...   
346   <a href="http://twitter.com/download/iphone" r...   
523   <a href="http://twitter.com/download/iphone" r...   
703   <a href="http://twitter.com/download/iphone" r...   
772   <a href="http://twitter.com/download/iphone" r...   
989   <a href="http://twitter.com/download/iphone" r...   
1723  <a href="https://about.twitter.com/products/tw...   
2086  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
56    Here is a pupper approaching maximum borkdrive...                14   
191   Here's a puppo participating in the #ScienceMa...                13   
192   I HEARD HE TIED HIS OWN BOWTIE MARK AND HE JUS...                13   
294   This is Waffles. He's a ship captain in real l...                13   
346                              @UNC can confirm 12/10                12   
523   I call this one "A Blep by the Sea" 12/10 http...                12   
703   This is Lucy. She's strives to be the best pot...                12   
772   This is Huck. He's addicted to caffeine. Hope ...                11   
989   Say hello to Divine Doggo. Must be magical af....                13   
1723  This pupper is not coming inside until she cat...                11   
2086  I would do radical things in the name of Dog G...                10   

      rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  \
56                    10        a      0        0       1      0    pupper   
191                   10        0      1        0       0      1     doggo   
192                   10        0      0        0       0      0     floof   
294                   10  Waffles      0        0       0      0     floof   
346                   10        0      0        0       0      0     floof   
523                   10        0      0        0       0      0     floof   
703                   10     Lucy      0        0       0      0     floof   
772                   10     Huck      0        0       1      0    pupper   
989                   10   Divine      1        0       0      0     doggo   
1723                  10        0      0        0       1      0    pupper   
2086                  10        0      0        0       0      0     floof   

      ratings_percentage  
56                   1.4  
191                  1.3  
192                  1.3  
294                  1.3  
346                  1.2  
523                  1.2  
703                  1.2  
772                  1.1  
989                  1.3  
1723                 1.1  
2086                 1.0

In [55]:
# looks good, but lets double check!
# range check: [54:56, 187:190, 289:291, 339:341, 515:517, 694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [55, 188, 189, 290, 340, 516, 695, 763, 979, 1712, 2074]
print("row 55 is gone")
test_filtered1.iloc[54:56]

row 55 is gone


tweet_id                  timestamp  \
54  881666595344535552  2017-07-03 00:11:11 +0000   
56  881536004380872706  2017-07-02 15:32:16 +0000   

                                               source  \
54  <a href="http://twitter.com/download/iphone" r...   
56  <a href="http://twitter.com/download/iphone" r...   

                                                 text  rating_numerator  \
54  This is Gary. He couldn't miss this puppertuni...                13   
56  Here is a pupper approaching maximum borkdrive...                14   

    rating_denominator  name  doggo  floofer  pupper  puppo Dog_Class  \
54                  10  Gary      0        0       0      0     floof   
56                  10     a      0        0       1      0    pupper   

    ratings_percentage  
54                 1.3  
56                 1.4

In [56]:
# range check: [187:190, 289:291, 339:341, 515:517, 694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [188, 189, 290, 340, 516, 695, 763, 979, 1712, 2074]
print("row 188 and 189 are gone")
test_filtered1.iloc[186:188]

row 188 and 189 are gone


tweet_id                  timestamp  \
187  856282028240666624  2017-04-23 23:01:59 +0000   
190  855857698524602368  2017-04-22 18:55:51 +0000   

                                                source  \
187  <a href="http://twitter.com/download/iphone" r...   
190  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
187  This is Cermet, Paesh, and Morple. They are ab...                14   
190  HE'S LIKE "WAIT A MINUTE I'M AN ANIMAL THIS IS...                13   

     rating_denominator    name  doggo  floofer  pupper  puppo Dog_Class  \
187                  10  Cermet      0        0       0      0     floof   
190                  10       0      0        0       0      0     floof   

     ratings_percentage  
187                 1.4  
190                 1.3

In [57]:
# range check: [289:291, 339:341, 515:517, 694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [290, 340, 516, 695, 763, 979, 1712, 2074]
print("row 290 is gone")
test_filtered1.iloc[286:288]

row 290 is gone


tweet_id                  timestamp  \
289  838201503651401729  2017-03-05 01:36:26 +0000   
291  838085839343206401  2017-03-04 17:56:49 +0000   

                                                source  \
289  <a href="http://twitter.com/download/iphone" r...   
291  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
289  RT @dog_rates: Meet Sunny. He can take down a ...                13   
291  @bragg6of8 @Andy_Pace_ we are still looking fo...                15   

     rating_denominator   name  doggo  floofer  pupper  puppo Dog_Class  \
289                  10  Sunny      0        0       0      0     floof   
291                  10      0      0        0       0      0     floof   

     ratings_percentage  
289                 1.3  
291                 1.5

In [58]:
# range check: [339:341, 515:517, 694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [340, 516, 695, 763, 979, 1712, 2074]
print("row 340 is gone")
test_filtered1.iloc[334:336]

row 340 is gone


tweet_id                  timestamp  \
339  832273440279240704  2017-02-16 17:00:25 +0000   
341  832215726631055365  2017-02-16 13:11:05 +0000   

                                                source  \
339  <a href="http://twitter.com/download/iphone" r...   
341  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
339  Say hello to Smiley. He's a blind therapy dogg...                14   
341  RT @dog_rates: This is Moreton. He's the Good ...                13   

     rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  \
339                  10   Smiley      1        0       0      0     doggo   
341                  10  Moreton      0        0       0      0     floof   

     ratings_percentage  
339                 1.4  
341                 1.3

In [59]:
# range check: [515:517, 694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [516, 695, 763, 979, 1712, 2074]
print("row 516 is gone")
test_filtered1.iloc[509:511]

row 516 is gone


tweet_id                  timestamp  \
515  811386762094317568  2016-12-21 01:44:13 +0000   
517  810896069567610880  2016-12-19 17:14:23 +0000   

                                                source  \
515  <a href="http://twitter.com/download/iphone" r...   
517  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
515  This is Craig. That's actually a normal sized ...                11   
517  This is Hunter. He just found out he needs bra...                11   

     rating_denominator    name  doggo  floofer  pupper  puppo Dog_Class  \
515                  10   Craig      0        0       1      0    pupper   
517                  10  Hunter      0        0       0      0     floof   

     ratings_percentage  
515                 1.1  
517                 1.1

In [60]:
# range check: [694:696, 762:764, 978:980, 1711:1713, 2073:2075]
# values removed [695, 763, 979, 1712, 2074]
print("row 695 is gone")
test_filtered1.iloc[687:689]

row 695 is gone


tweet_id                  timestamp  \
694  786729988674449408  2016-10-14 00:47:00 +0000   
696  786664955043049472  2016-10-13 20:28:35 +0000   

                                                source  \
694  <a href="http://twitter.com/download/iphone" r...   
696  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
694  RT @dog_rates: This is Oakley. He has no idea ...                11   
696  "Honestly Kathleen I just want more Ken Bone" ...                12   

     rating_denominator    name  doggo  floofer  pupper  puppo Dog_Class  \
694                  10  Oakley      0        0       0      0     floof   
696                  10       0      0        0       0      0     floof   

     ratings_percentage  
694                 1.1  
696                 1.2

In [61]:
# range check: [762:764, 978:980, 1711:1713, 2073:2075]
# values removed [763, 979, 1712, 2074]
print("row 763 is gone")
test_filtered1.iloc[754:756]

row 763 is gone


tweet_id                  timestamp  \
762  778039087836069888  2016-09-20 01:12:28 +0000   
764  777953400541634568  2016-09-19 19:31:59 +0000   

                                                source  \
762  <a href="http://twitter.com/download/iphone" r...   
764  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
762  Evolution of a pupper yawn featuring Max. 12/1...                12   
764  RT @dog_rates: Meet Gerald. He's a fairly exot...                 8   

     rating_denominator    name  doggo  floofer  pupper  puppo Dog_Class  \
762                  10       0      0        0       1      0    pupper   
764                  10  Gerald      1        0       0      0     doggo   

     ratings_percentage  
762                 1.2  
764                 0.8

In [62]:
# range check: [978:980, 1711:1713, 2073:2075]
# values removed [979, 1712, 2074]
print("row 979 is gone")
test_filtered1.iloc[969:971]

row 979 is gone


tweet_id                  timestamp  \
978  749996283729883136  2016-07-04 16:00:22 +0000   
980  749774190421639168  2016-07-04 01:17:51 +0000   

                                                source  \
978  <a href="https://about.twitter.com/products/tw...   
980  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
978  This is Bo. He emanates happiness. 12/10 I cou...                12   
980  This is Lucy. She's a Benebop Cumberplop. 12/1...                12   

     rating_denominator  name  doggo  floofer  pupper  puppo Dog_Class  \
978                  10    Bo      0        0       0      0     floof   
980                  10  Lucy      0        0       0      0     floof   

     ratings_percentage  
978                 1.2  
980                 1.2

In [63]:
# range check: [1711:1713, 2073:2075]
# values removed [1712, 2074]
print("row 1712 is gone")
test_filtered1.iloc[1701:1703]

row 1712 is gone


tweet_id                  timestamp  \
1711  680497766108381184  2015-12-25 21:18:05 +0000   
1713  680473011644985345  2015-12-25 19:39:43 +0000   

                                                 source  \
1711  <a href="http://twitter.com/download/iphone" r...   
1713  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1711  Meet Percy. He's a Latvian Yuletide Heineken. ...                12   
1713  This is Hector. He thinks he's a hammer. Silly...                10   

      rating_denominator    name  doggo  floofer  pupper  puppo Dog_Class  \
1711                  10   Percy      0        0       0      0     floof   
1713                  10  Hector      0        0       1      0    pupper   

      ratings_percentage  
1711                 1.2  
1713                 1.0

In [64]:
# range check: [2073:2075]
# values removed [2074]
print("row 2074 is gone")
test_filtered1.iloc[2062:2064]

row 2074 is gone


tweet_id                  timestamp  \
2073  670995969505435648  2015-11-29 16:01:20 +0000   
2075  670840546554966016  2015-11-29 05:43:44 +0000   

                                                 source  \
2073  <a href="http://twitter.com/download/iphone" r...   
2075  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
2073  Yea I can't handle this job anymore your dogs ...                12   
2075  Meet Colby. He's that one cool friend that get...                10   

      rating_denominator   name  doggo  floofer  pupper  puppo Dog_Class  \
2073                  10      0      0        0       0      0     floof   
2075                  10  Colby      0        0       0      0     floof   

      ratings_percentage  
2073                 1.2  
2075                 1.0

✔️ Ok, we are good! The columns we wanted gone are gone!

In [65]:
# reassign test_filtered1 to dfc_archive again
print("before dataframe conversion = ", test_filtered1.shape)
dfc_archive = test_filtered1
print("after dataframe conversion = ", dfc_archive.shape)
dfc_archive.head(5)

before dataframe conversion =  (2344, 13)
after dataframe conversion =  (2344, 13)


tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   
3  This is Darla. She commenced a snooze mid meal...                13   
4  This is Franklin. He would like you to stop ca...                12   

   rating_denominator      name  doggo  floofer  pupper  puppo Dog_Class  \
0                  10   Phineas      0        0       0      0     floof   
1                  10     Tilly      0        0       0      0     floof   
2                  10    Archie      0        0       0      0     floof   
3                  10     Darla      0        0       0      0     floof   
4                  10  Franklin      0        0       0      0     floof   

   ratings_percentage  
0                 1.3  
1                 1.3  
2                 1.2  
3                 1.3  
4                 1.2

---

<a id='Q7'></a>

<font color='red'>


### ✔️ 7. df_archive --> remove entry 313 because no denominator is provided (entered as 0)
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [66]:
# create test df for prototyping
test = dfc_archive

test.iloc[308:310]

tweet_id                  timestamp  \
312  835264098648616962  2017-02-24 23:04:14 +0000   
314  835172783151792128  2017-02-24 17:01:22 +0000   

                                                source  \
312  <a href="http://twitter.com/download/iphone" r...   
314  <a href="http://twitter.com/download/iphone" r...   

                                                  text  rating_numerator  \
312  Meet Lola. Her hobbies include being precious ...                12   
314  We only rate dogs. Please don't send in any no...                12   

     rating_denominator  name  doggo  floofer  pupper  puppo Dog_Class  \
312                  10  Lola      0        0       0      0     floof   
314                  10     0      0        0       0      0     floof   

     ratings_percentage  
312                 1.2  
314                 1.2

<a id='scope'></a>
Please note that this "inf" value or "zero denominator" was removed earlier during cleaning process (automatically). Visual confirmation confirmed above!

<a href="#Q7_solved">[Go to Quality Cleaning #7]</a>

---

<a id='Q8'></a>

<font color='red'>

### ✔️ 8. df_prediction --> make all names of dogs CAPITALIZED, to prevent any variants of different capitilizations
    
<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [67]:
test = dfc_prediction
test.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

In [68]:
# checking if "pX_dog" are boolean - we are good, just make "pX" of the three columns all caps!
test.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

In [69]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [70]:
# help: https://www.geeksforgeeks.org/apply-uppercase-to-a-column-in-pandas-dataframe/
test['p1'] = test['p1'].str.upper()
test['p2'] = test['p2'].str.upper()
test['p3'] = test['p3'].str.upper()
test.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  WELSH_SPRINGER_SPANIEL  0.465074    True              COLLIE   
1        1                 REDBONE  0.506826    True  MINIATURE_PINSCHER   
2        1         GERMAN_SHEPHERD  0.596461    True            MALINOIS   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    SHETLAND_SHEEPDOG  0.061428    True  
1  0.074192    True  RHODESIAN_RIDGEBACK  0.072010    True  
2  0.138584    True           BLOODHOUND  0.116197    True

In [71]:
# assign test back to orignal dataframe
dfc_prediction = test
dfc_prediction.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  WELSH_SPRINGER_SPANIEL  0.465074    True              COLLIE   
1        1                 REDBONE  0.506826    True  MINIATURE_PINSCHER   
2        1         GERMAN_SHEPHERD  0.596461    True            MALINOIS   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    SHETLAND_SHEEPDOG  0.061428    True  
1  0.074192    True  RHODESIAN_RIDGEBACK  0.072010    True  
2  0.138584    True           BLOODHOUND  0.116197    True

In [72]:
# make names in dfc_archive all caps too
test = dfc_archive
test.head(3)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   

   rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  \
0                  10  Phineas      0        0       0      0     floof   
1                  10    Tilly      0        0       0      0     floof   
2                  10   Archie      0        0       0      0     floof   

   ratings_percentage  
0                 1.3  
1                 1.3  
2                 1.2

In [73]:
# need to remove warning... but is ok for now
# help: https://www.dataquest.io/blog/settingwithcopywarning/
test['name'] = test['name'].str.upper()
test.head(3)

<ipython-input-73-86b1997fe849>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test['name'] = test['name'].str.upper()


tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   

   rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  \
0                  10  PHINEAS      0        0       0      0     floof   
1                  10    TILLY      0        0       0      0     floof   
2                  10   ARCHIE      0        0       0      0     floof   

   ratings_percentage  
0                 1.3  
1                 1.3  
2                 1.2

In [74]:
# assign test back to orignal dataframe
dfc_archive = test
dfc_archive.head(3)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   

   rating_denominator     name  doggo  floofer  pupper  puppo Dog_Class  \
0                  10  PHINEAS      0        0       0      0     floof   
1                  10    TILLY      0        0       0      0     floof   
2                  10   ARCHIE      0        0       0      0     floof   

   ratings_percentage  
0                 1.3  
1                 1.3  
2                 1.2

---

<a id='tidy-section'></a>

<font color='red'>

## TIDINESS ISSUES ADDRESSED -- note that ```df``` will change to ```dfc``` to indicate cleaned data

<a href="#clean">[Cleaning Data Requirements]</a> 
<a href="#todo">[Cleaning and Tidying To-do List]</a> 
<a href="#contents">[Table of Contents]</a>

<a id='T1'></a>

<font color='red'>

### 1. Combine all seperated tables, join by ```tweet_id```

<a href="#todo">[Cleaning and Tidying To-do List]</a>

<a id='T2'></a>

<font color='red'>

### 2. df_twitter drop column ```full_text``` because it contains multiple observational units (e.g., name, url, score, etc.)

<a href="#todo">[Cleaning and Tidying To-do List]</a>

<hr size="5"/>

## Cleaning Specification 3.2 - Output clean data as "tidy master dataset"

**3.2 CRITERIA:** The student is able to thoroughly clean a dataset programmatically.

**3.2 SPECIFICATION:**

Copies of the original pieces of data are made prior to cleaning.

All issues identified in the assess phase are successfully cleaned (if possible) using Python and pandas, and include the cleaning tasks required to satisfy the Project Motivation.

A tidy master dataset (or datasets, if appropriate) with all pieces of gathered data is created.
    
<a href="#clean">[Cleaning Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

<hr size="5"/>

<a id='store'></a>
# 4. STORING AND ACTING ON WRANGLED DATA

<a href="#contents">[Table of Contents]</a>

<font color=blue>
    
**4.1 CRITERIA:** The student is able to store a gathered, assessed, and cleaned dataset.

**4.1 SPECIFICATION:**

Students will save their gathered, assessed, and cleaned master dataset(s) to a CSV file or a SQLite database.

<font color=blue>
    
**4.2 CRITERIA:** The student is able to act on their wrangled data to produce insights (e.g. analyses, visualizations, and/or models).

**4.2 SPECIFICATION:**

The master dataset is analyzed using pandas or SQL in the Jupyter Notebook and at least three (3) separate insights are produced.

At least one (1) labeled visualization is produced in the Jupyter Notebook using Python’s plotting libraries or in Tableau.

Students must make it clear in their wrangling work that they assessed and cleaned (if necessary) the data upon which the analyses and visualizations are based.

<hr size="5"/>

<a id='report'></a>
# 5. REPORT-DISCUSSION-CONCLUSION

<a href="#contents">[Table of Contents]</a>

<font color=blue>
    
**5.1 CRITERIA:** The student is able to reflect upon and describe their data wrangling efforts.

**5.1 SPECIFICATION:**

The student’s wrangling efforts are briefly described. This document (wrangle_report.pdf or wrangle_report.html) is concise and approximately 300-600 words in length.

At least one (1) labeled visualization is produced in the Jupyter Notebook using Python’s plotting libraries or in Tableau.

Students must make it clear in their wrangling work that they assessed and cleaned (if necessary) the data upon which the analyses and visualizations are based.

<font color=blue>
    
**5.2 CRITERIA:** The student is able to describe some insights found in their wrangled dataset.

**5.2 SPECIFICATION:**

The three (3) or more insights the student found are communicated. At least one (1) visualization is included.

This document (act_report.pdf or act_report.html) is at least 250 words in length.

<hr size="5"/>

<a id='files'></a>
# 6. PROJECT FILES

<a href="#contents">[Table of Contents]</a>

<font color=blue>
    
**6.1 CRITERIA:** Are all required files included in the student's submission?

**6.1 SPECIFICATION:**
The following files (with identical filenames) are included:

* wrangle_act.ipynb
* wrangle_report.pdf or wrangle_report.html
* act_report.pdf or act_report.html
    
All dataset files are included, including the stored master dataset(s), with filenames and extensions as specified on the Project Submission page.

<hr size="5"/>

# End of Data Project!

Made with ❤️ by Jhon!

Further impovements include...
1. Validation statements using ```assert``` clauses to confirm that data manipuation is correct during cleaning stage

<a href="#contents">[Table of Contents]</a>